### ISSUE TITLE
Some ctor-like class method doen't give the correct object type

### ISSUE THREAD

**Issue by bew** [crystal-lang/crystal#5335]
I know inheriting from `File` is probably not a good idea, but it shows something which looks like a bug in the stdlib:
```cr
class MyFile < File
end

# Without a block
puts typeof(MyFile.open "/tmp")

# With a block
MyFile.open "/tmp" do |f|
  puts typeof(f)
end
```
Gives:
```
MyFile
File
```

I would expect to get a `MyFile` all the time.

This happens because in the implementation:
- `File.open` without a block: uses `new` to create the file object
- `File.open` with a block: uses `File.new` to create the file object

The first solution I see would be to remove all `File.new` from `File` constructor-like class methods (like `open`) and use `new` instead.

But there is another issue with this solution: This will _work until I add a new constructor_ to `MyType`, for example taking no arguments, and manually initializing the class fields.
Because when I do that all the other constructors are discarded, and the one used in `File.open` doesn't exists anymore unless I re-define it in `MyFile`.
However I'm not sure my concern about discarded constructor is really a problem actually.

I noticed it for `File` but this kind of issue probably exists for other types too.

**Comment by ysbaddaden** [comment]
I'm not sure this is a bug. You call an ancestor class method that yields self, and AFAIK the ancestor method isn't duplicated for each subclass but implemented once, so self can only statically refer to a File, not a MyFile (which is correct because MyFile extends File).

**Comment by bew** [comment]
I'm not sure, as this works:
```cr
class A
  def self.build
    instance = new
    yield instance
  end

  def self.build_a
    instance = A.new
    yield instance
  end
end

class B < A
end

B.build do |i|
  pp typeof(i)   # => B
end

B.build_a do |i_a|
  pp typeof(i_a) # => A
end
```
If you use `new` instead of `A.new` it builds the correct instance and yield it!

**Comment by ysbaddaden** [comment]
You never call `A.build`. Does Crystal generates `B.build` (that instantiates B) instead of `A.build` (that would likely instanciate A)?

**Comment by bew** [comment]
I'm not sure I get your comment, but checking every cases it gives the correct type (in my sense) everytime, based on which ctor it called (https://carc.in/#/r/369g):
```cr
class A
  def self.build
    instance = new
    yield instance
  end

  def self.build_a
    instance = A.new
    yield instance
  end
end

class B < A
end

A.build do |i|
  pp typeof(i)   # => A
end

A.build_a do |i_a|
  pp typeof(i_a) # => A
end

B.build do |i|
  pp typeof(i)   # => B
end

B.build_a do |i_a|
  pp typeof(i_a) # => A
end
```

**Comment by larubujo** [comment]
thats why inheritance is bad. if go, no inheritance, problems and headache solved. dont inherit file. code smell

**Comment by javanut13** [comment]
@larubujo As Crystal is an object oriented language which supports inheritance, even if situations like this could be considered a code smell, it is important that they work as a user would expect. Having unpredictable behaviour - for something as common as inheritance - where the solution is “don’t do that” will frustrate new developers.

**Comment by larubujo** [comment]
@javanut13 inheriting file is nothing but wrong. composition should be used. in fact inheritance very little use case. remove inheritance, like in go, problem solved. making it work means more code, more effort, more edge cases, all because one developer wanted to inherit instead of compose. op should explain why need inherit file.

**Comment by javanut13** [comment]
@larubujo I don’t think this is the place to debate inheritance vs composition. Go is a very different language to Crystal, but if you want to propose this change then it’s probably best to discuss it on Glitter or open a new issue outlining the trade offs of each potential solution.

**Comment by RX14** [comment]
@larubujo we're not going to remove inheritance, regardless of anyone's opinions on it because it's simply too far along in the language's development to make these decisions. Not every programming language will meet a person's tastes, please keep your comments constructive.

**Comment by Sija** [comment]
Should this issue be closed, since #5337 was merged?

**Comment by bew** [comment]
I think this can be closed, as I currently didn't find another place where this might be relevant. Feel free to re-open if you find some more.
